File: python/tests/test_example_c.py
"""
Validation tests for Example C (inverted pendulum, numerical predictor).
Ponomarev (2016), Section VI-C, Eq. 89-104.
This is the main example, reproducing Fig. 1 of the paper.
"""

import sys
import os
import numpy as np
import pytest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from src.example_c import run_example_c
from src.uncompensated import run_uncompensated


class TestExampleCConvergenceAllICs:
    """test_convergence_all_ics -- Fig. 1: three initial conditions all converge."""

    @pytest.mark.parametrize("x1_init", [np.pi / 2, np.pi, 3 * np.pi / 2])
    def test_converges(self, x1_init):
        """State should converge for each initial condition."""
        params = {'x0': [x1_init, np.pi / 2.0], 'u0': 1.0}
        res = run_example_c(params)
        final_norm = np.linalg.norm(res['x_hist'][-1])
        print(f'    x1(0) = {x1_init:.4f}: |x(t_end)| = {final_norm:.4e}')
        # Euler at dt=0.01 produces residuals ~0.03-0.07; use 0.1 threshold
        assert final_norm < 0.1, \
            f'State did not converge for x1(0) = {x1_init:.4f}.'


class TestExampleCInitialControlSpike:
    """test_initial_control_spike -- First feedback value has large magnitude."""

    def test_large_spike(self):
        """Control should exhibit a large initial spike (per Fig. 1)."""
        params = {'x0': [np.pi, np.pi / 2.0], 'u0': 1.0}
        res = run_example_c(params)

        u_first = abs(res['u_hist'][0])
        u_max   = np.max(np.abs(res['u_hist']))
        print(f'    |u(1)| = {u_first:.4f}')
        print(f'    max|u| = {u_max:.4f}')

        assert u_max > 5, \
            'Control should exhibit a large initial spike (per Fig. 1).'


class TestExampleCPredictorAtT0:
    """test_predictor_at_t0 -- Verify predictor ODE at t=0 vs independent integration."""

    def test_predictor_matches(self):
        """Predictor at t=0 should match independent forward integration."""
        h  = np.pi / 4.0
        x0 = np.array([np.pi, np.pi / 2.0])
        u0 = 1.0
        dt = 0.01
        N_h = int(round(h / dt))
        ds = h / N_h

        # Manually solve predictor ODE at t=0
        xi = x0.copy()
        for j in range(N_h):
            dxi = np.array([xi[1], np.sin(xi[0]) + u0])
            xi = xi + ds * dxi
        y_manual = xi.copy()

        # Run full simulation and extract y(0)
        params = {'x0': x0.tolist(), 'u0': u0, 'h': h, 'dt': dt}
        res = run_example_c(params)
        y_sim = res['y_hist'][0, :]

        err = np.linalg.norm(y_manual - y_sim)
        print(f'    y_manual = [{y_manual[0]:.6f}, {y_manual[1]:.6f}]')
        print(f'    y_sim    = [{y_sim[0]:.6f}, {y_sim[1]:.6f}]')
        print(f'    |error|  = {err:.4e}')

        assert err < 1e-10, \
            'Predictor at t=0 should match independent forward integration.'


class TestExampleCCompensatedVsUncompensated:
    """test_compensated_vs_uncompensated -- Predictor converges, open-loop diverges."""

    def test_compensated_converges(self):
        """Compensated system should converge."""
        x0 = [np.pi, np.pi / 2.0]
        params_comp = {'x0': x0, 'u0': 1.0}
        res_comp = run_example_c(params_comp)
        norm_comp = np.linalg.norm(res_comp['x_hist'][-1])
        print(f'    Compensated:   |x(t_end)| = {norm_comp:.4e}')
        # Euler at dt=0.01 produces residuals ~0.04; use 0.1 threshold
        assert norm_comp < 0.1, 'Compensated system should converge.'

    def test_uncompensated_worse(self):
        """Uncompensated should perform worse than compensated."""
        x0 = [np.pi, np.pi / 2.0]

        params_comp = {'x0': x0, 'u0': 1.0}
        res_comp = run_example_c(params_comp)
        norm_comp = np.linalg.norm(res_comp['x_hist'][-1])

        params_unc = {'x0': x0, 'u0': 1.0, 'control_mode': 'zero'}
        res_unc = run_uncompensated(params_unc)
        norm_unc = np.linalg.norm(res_unc['x_hist'][-1])

        print(f'    Compensated:   |x(t_end)| = {norm_comp:.4e}')
        print(f'    Uncompensated: |x(t_end)| = {norm_unc:.4e}')
        assert norm_unc > norm_comp, \
            'Uncompensated should perform worse than compensated.'


class TestExampleCDtRefinement:
    """test_dt_refinement -- O(dt) convergence order estimate."""

    def test_convergence_order(self):
        """Average convergence order should be at least 0.8."""
        params_base = {'x0': [np.pi, np.pi / 2.0], 'u0': 1.0, 't_end': 5.0}

        dt_vals = [0.04, 0.02, 0.01, 0.005]

        # Reference: dt = 0.0025
        params_ref = {**params_base, 'dt': 0.0025}
        res_ref = run_example_c(params_ref)

        errors = np.zeros(len(dt_vals))
        for i, dt_i in enumerate(dt_vals):
            params_i = {**params_base, 'dt': dt_i}
            res_i = run_example_c(params_i)

            # Interpolate reference onto this grid
            x_ref_interp = np.zeros_like(res_i['x_hist'])
            for col in range(2):
                x_ref_interp[:, col] = np.interp(res_i['t'], res_ref['t'],
                                                   res_ref['x_hist'][:, col])
            errors[i] = np.max(np.abs(res_i['x_hist'] - x_ref_interp))
            print(f'    dt = {dt_i:.4f}: max error = {errors[i]:.4e}')

        # Compute convergence orders between successive refinements
        print('    Convergence orders:')
        orders = []
        for i in range(len(dt_vals) - 1):
            if errors[i + 1] > 1e-14:
                ratio = dt_vals[i] / dt_vals[i + 1]
                order = np.log(errors[i] / errors[i + 1]) / np.log(ratio)
                orders.append(order)
                print(f'      dt {dt_vals[i]:.4f} -> {dt_vals[i+1]:.4f}: '
                      f'order = {order:.2f}')

        if orders:
            avg_order = np.mean(orders)
            print(f'    Average order = {avg_order:.2f} (expected ~1.0)')
            assert avg_order > 0.8, \
                'Average convergence order should be at least 0.8.'


class TestExampleCLyapunovV0:
    """test_lyapunov_v0 -- ||y||^2 should decrease after transient."""

    def test_y_norm_decays(self):
        """||y||^2 should decay to less than 10% of post-transient value."""
        params = {'x0': [np.pi, np.pi / 2.0], 'u0': 1.0}
        res = run_example_c(params)

        h = np.pi / 4.0  # default delay
        y_norm_sq = np.sum(res['y_hist']**2, axis=1)

        idx_start = np.searchsorted(res['t'], 2.0 * h)
        y_tail = y_norm_sq[idx_start:]

        ratio = y_tail[-1] / y_tail[0]
        print(f'    ||y(t_end)||^2 / ||y(2h)||^2 = {ratio:.4e}')
        assert ratio < 0.1, \
            '||y||^2 should decay to less than 10% of post-transient value.'

    def test_mostly_non_increasing(self):
        """||y||^2 should be mostly non-increasing after transient."""
        params = {'x0': [np.pi, np.pi / 2.0], 'u0': 1.0}
        res = run_example_c(params)

        h = np.pi / 4.0
        y_norm_sq = np.sum(res['y_hist']**2, axis=1)

        idx_start = np.searchsorted(res['t'], 2.0 * h)
        y_tail = y_norm_sq[idx_start:]

        dV = np.diff(y_tail)
        frac_dec = np.sum(dV <= 1e-6) / len(dV)
        print(f'    Fraction non-increasing (t > 2h): {frac_dec*100:.2f}%')
        assert frac_dec > 0.80, \
            '||y||^2 should be mostly non-increasing after transient.'


class TestExampleCDelaySweep:
    """test_delay_sweep -- Multiple delay values all converge."""

    @pytest.mark.parametrize("h_val", [0.5, np.pi / 4, 1.0])
    def test_converges(self, h_val):
        """State should converge for each delay value."""
        params = {'x0': [np.pi, np.pi / 2.0], 'u0': 1.0,
                  'h': h_val, 't_end': 15.0}
        res = run_example_c(params)
        final_norm = np.linalg.norm(res['x_hist'][-1])
        print(f'    h = {h_val:.4f}: |x(t_end)| = {final_norm:.4e}')
        assert final_norm < 0.1, \
            f'State should converge for h = {h_val:.4f}.'

File: matlab/tests/test_example_c.m
function tests = test_example_c
%TEST_EXAMPLE_C  Validation tests for Example C (inverted pendulum, numerical predictor).
%
%  Ponomarev (2016), Section VI-C, Eq. 89-104.
%  This is the main example, reproducing Fig. 1 of the paper.
%
%  Tests:
%    test_convergence_all_ics       — x1(0) in {pi/2, pi, 3pi/2} all converge
%    test_initial_control_spike     — |u(1)| > 5 per Fig. 1
%    test_predictor_at_t0           — predictor ODE at t=0 matches independent integration
%    test_compensated_vs_uncompensated — predictor converges, open-loop diverges
%    test_dt_refinement             — O(dt) convergence order estimate
%    test_lyapunov_v0               — ||y||^2 decreases after transient
%    test_delay_sweep               — multiple delay values all converge
    tests = functiontests(localfunctions);
end

%% test_convergence_all_ics — Fig. 1: three initial conditions all converge
function test_convergence_all_ics(testCase)
    fprintf('\n  [Example C] test_convergence_all_ics: 3 initial conditions\n');

    x1_inits = [pi/2, pi, 3*pi/2];
    x2_init  = pi/2;

    for i = 1:length(x1_inits)
        params = struct('x0', [x1_inits(i); x2_init], 'u0', 1.0);
        [t, x_hist, u_hist] = run_example_c(params);

        final_norm = norm(x_hist(end,:));
        fprintf('    x1(0) = %.4f: |x(t_end)| = %.4e\n', x1_inits(i), final_norm);
        verifyLessThan(testCase, final_norm, 0.1, ...
            sprintf('State did not converge for x1(0) = %.4f.', x1_inits(i)));
    end
end

%% test_initial_control_spike — First feedback value has large magnitude
function test_initial_control_spike(testCase)
    fprintf('\n  [Example C] test_initial_control_spike: |u(1)| > 5\n');

    % Use x1(0) = pi (the middle initial condition from Fig. 1)
    params = struct('x0', [pi; pi/2], 'u0', 1.0);
    [~, ~, u_hist] = run_example_c(params);

    % The first feedback value (index 1 gives the first control applied)
    % Paper Fig. 1 shows initial spike u ~ -8 to -9
    u_first = abs(u_hist(1));
    u_max   = max(abs(u_hist));
    fprintf('    |u(1)| = %.4f\n', u_first);
    fprintf('    max|u| = %.4f\n', u_max);

    verifyGreaterThan(testCase, u_max, 5, ...
        'Control should exhibit a large initial spike (per Fig. 1).');
end

%% test_predictor_at_t0 — Verify predictor ODE at t=0 vs independent integration
function test_predictor_at_t0(testCase)
    fprintf('\n  [Example C] test_predictor_at_t0: independent predictor check\n');

    h  = pi/4;
    x0 = [pi; pi/2];
    u0 = 1.0;
    dt = 0.01;
    N_h = round(h / dt);
    ds = h / N_h;

    % Manually solve the predictor ODE at t=0:
    % xi'(s) = [xi2; sin(xi1) + u(0 + s - h)]
    % xi(0) = x0 = [pi; pi/2]
    % For s in [0, h]: u(s - h) = u0 = 1.0 (within pre-history)
    xi = x0;
    for j = 0:N_h-1
        dxi = [xi(2); sin(xi(1)) + u0];
        xi = xi + ds * dxi;
    end
    y_manual = xi;

    % Run full simulation and extract y(0) = y_hist(1,:)
    params = struct('x0', x0, 'u0', u0, 'h', h, 'dt', dt);
    [~, ~, ~, y_hist] = run_example_c(params);
    y_sim = y_hist(1,:)';

    err = norm(y_manual - y_sim);
    fprintf('    y_manual = [%.6f, %.6f]\n', y_manual(1), y_manual(2));
    fprintf('    y_sim    = [%.6f, %.6f]\n', y_sim(1), y_sim(2));
    fprintf('    |error|  = %.4e\n', err);

    verifyLessThan(testCase, err, 1e-10, ...
        'Predictor at t=0 should match independent forward integration.');
end

%% test_compensated_vs_uncompensated — Predictor converges, open-loop diverges
function test_compensated_vs_uncompensated(testCase)
    fprintf('\n  [Example C] test_compensated_vs_uncompensated\n');

    x0 = [pi; pi/2];

    % Compensated (predictor feedback)
    params_comp = struct('x0', x0, 'u0', 1.0);
    [~, x_comp, ~] = run_example_c(params_comp);
    norm_comp = norm(x_comp(end,:));

    % Uncompensated (open-loop, u=0)
    params_unc = struct('x0', x0, 'u0', 1.0, 'control_mode', 'zero');
    [~, x_unc, ~] = run_uncompensated(params_unc);
    norm_unc = norm(x_unc(end,:));

    fprintf('    Compensated:   |x(t_end)| = %.4e\n', norm_comp);
    fprintf('    Uncompensated: |x(t_end)| = %.4e\n', norm_unc);

    % Compensated should converge (Euler at dt=0.01 gives residual ~0.04)
    verifyLessThan(testCase, norm_comp, 0.1, ...
        'Compensated system should converge.');

    % Uncompensated should fail to converge (or diverge)
    % The inverted pendulum at x0 = [pi, pi/2] with u=0 will not converge
    verifyGreaterThan(testCase, norm_unc, norm_comp, ...
        'Uncompensated should perform worse than compensated.');
end

%% test_dt_refinement — O(dt) convergence order estimate
function test_dt_refinement(testCase)
    fprintf('\n  [Example C] test_dt_refinement: convergence order\n');

    params_base = struct('x0', [pi; pi/2], 'u0', 1.0, 't_end', 5.0);

    % Multiple dt values
    dt_vals = [0.04, 0.02, 0.01, 0.005];

    % Reference: dt = 0.0025
    params_ref = params_base;
    params_ref.dt = 0.0025;
    [t_ref, x_ref, ~] = run_example_c(params_ref);

    errors = zeros(size(dt_vals));
    for i = 1:length(dt_vals)
        params_i = params_base;
        params_i.dt = dt_vals(i);
        [t_i, x_i, ~] = run_example_c(params_i);

        % Interpolate reference onto this grid
        x_ref_interp = interp1(t_ref, x_ref, t_i, 'linear');
        errors(i) = max(max(abs(x_i - x_ref_interp)));
        fprintf('    dt = %.4f: max error = %.4e\n', dt_vals(i), errors(i));
    end

    % Compute convergence orders between successive refinements
    fprintf('    Convergence orders:\n');
    orders = zeros(length(dt_vals)-1, 1);
    for i = 1:length(dt_vals)-1
        if errors(i+1) > 1e-14
            ratio = dt_vals(i) / dt_vals(i+1);
            orders(i) = log(errors(i) / errors(i+1)) / log(ratio);
            fprintf('      dt %.4f -> %.4f: order = %.2f\n', ...
                dt_vals(i), dt_vals(i+1), orders(i));
        end
    end

    % Average order should be close to 1.0 (Euler method)
    valid_orders = orders(orders > 0);
    if ~isempty(valid_orders)
        avg_order = mean(valid_orders);
        fprintf('    Average order = %.2f (expected ~1.0)\n', avg_order);
        verifyGreaterThan(testCase, avg_order, 0.8, ...
            'Average convergence order should be at least 0.8.');
    end
end

%% test_lyapunov_v0 — ||y||^2 should decrease after transient
function test_lyapunov_v0(testCase)
    fprintf('\n  [Example C] test_lyapunov_v0: ||y||^2 decay\n');

    params = struct('x0', [pi; pi/2], 'u0', 1.0);
    [t, ~, ~, y_hist] = run_example_c(params);

    h = pi/4;  % default delay

    % Compute ||y||^2 at each time step
    y_norm_sq = sum(y_hist.^2, 2);

    % After transient (t > 2*h), check that ||y||^2 generally decreases
    idx_start = find(t >= 2*h, 1, 'first');
    y_tail = y_norm_sq(idx_start:end);

    % Check that final value is much smaller than post-transient start
    ratio = y_tail(end) / y_tail(1);
    fprintf('    ||y(t_end)||^2 / ||y(2h)||^2 = %.4e\n', ratio);
    verifyLessThan(testCase, ratio, 0.1, ...
        '||y||^2 should decay to less than 10%% of post-transient value.');

    % Count fraction of non-increasing steps
    dV = diff(y_tail);
    frac_dec = sum(dV <= 1e-6) / length(dV);
    fprintf('    Fraction non-increasing (t > 2h): %.2f%%\n', frac_dec*100);
    % Be lenient: Euler can cause small oscillations
    verifyGreaterThan(testCase, frac_dec, 0.80, ...
        '||y||^2 should be mostly non-increasing after transient.');
end

%% test_delay_sweep — Multiple delay values all converge
function test_delay_sweep(testCase)
    fprintf('\n  [Example C] test_delay_sweep: h in {0.2, 0.5, 0.7854, 1.0}\n');

    h_vals = [0.5, pi/4, 1.0];
    tol = 0.1;  % relaxed tolerance for Euler discretization

    for i = 1:length(h_vals)
        params = struct('x0', [pi; pi/2], 'u0', 1.0, 'h', h_vals(i), ...
                         't_end', 15.0);
        [~, x_hist, ~] = run_example_c(params);

        final_norm = norm(x_hist(end,:));
        fprintf('    h = %.4f: |x(t_end)| = %.4e\n', h_vals(i), final_norm);
        verifyLessThan(testCase, final_norm, tol, ...
            sprintf('State should converge for h = %.4f.', h_vals(i)));
    end
end
